sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -1248,20 +1307,44 @@ public void processElement(ProcessContext c) {
     }
   }
 
-  private static class WriteToSpannerFn extends DoFn<Iterable<MutationGroup>, Void> {
+  @VisibleForTesting
+  static class WriteToSpannerFn extends DoFn<Iterable<MutationGroup>, Void> {
 
     private transient SpannerAccessor spannerAccessor;
     private final SpannerConfig spannerConfig;
     private final FailureMode failureMode;
-    private final Counter mutationGroupBatchesCounter =
-        Metrics.counter(WriteGrouped.class, "mutation_group_batches");
-    private final Counter mutationGroupWriteSuccessCounter =
+
+    @VisibleForTesting static Sleeper sleeper = Sleeper.DEFAULT;
+
+    private final Counter mutationGroupBatchesReceived =
+        Metrics.counter(WriteGrouped.class, "mutation_group_batches_received");
+    private final Counter mutationGroupBatchesWriteSuccess =
+        Metrics.counter(WriteGrouped.class, "mutation_group_batches_write_success");
+    private final Counter mutationGroupBatchesWriteFail =
+        Metrics.counter(WriteGrouped.class, "mutation_group_batches_write_fail");
+
+    private final Counter mutationGroupsReceived =
+        Metrics.counter(WriteGrouped.class, "mutation_groups_received");
+    private final Counter mutationGroupsWriteSuccess =
         Metrics.counter(WriteGrouped.class, "mutation_groups_write_success");
-    private final Counter mutationGroupWriteFailCounter =
+    private final Counter mutationGroupsWriteFail =
         Metrics.counter(WriteGrouped.class, "mutation_groups_write_fail");
 
+    private final Counter spannerWriteSuccess =
+        Metrics.counter(WriteGrouped.class, "spanner_write_success");
+    private final Counter spannerWriteFail =
+        Metrics.counter(WriteGrouped.class, "spanner_write_fail");
+    private final Counter spannerWriteTotalLatency =
```

Comment:
Would an average writeLatency(if its possible to calculate) be more actionable to the user rather than a total write latency? In the sense, if the write latency is high then the user can reduce the workers.

Reply:
Its not possible - the only things we can export are counters and gauges.

User will have to manually calculate average latency.